Std/Data/Nat/Basic.lean

```diff
@@ -126,3 +126,92 @@ where
     else
       guess
 termination_by iter guess => guess
+
+/-- `bodd n` returns `true` if `n` is odd-/
+def bodd (n : Nat) : Bool :=
+  (1 &&& n) != 0
+
+/-- `div2 n = ⌊n/2⌋` the greatest integer smaller than `n/2`-/
+def div2 (n : Nat) : Nat :=
+  n >>> 1
+
+/-- `bit b` appends the digit `b` to the binary representation of
+  its natural number input. -/
+def bit (b : Bool) (n : Nat) : Nat :=
+  cond b (n + n + 1) (n + n)
```

Comment:
I'm not a huge fan of `bit` in this namespace as is.  The meaning isn't obvious to me and `Nat`.  Could we replace `bit b n` with `2*n + ofBool b` where `abbrev ofBool b := cond b 1 0`?

Reply:
I have lesser, but similar concerns about `div2`.  Perhaps just write `n / 2`?

Reply:
Agreed, these definitions are holdovers from the lean 3 way of representing numbers, and I think we can just do without them. `bit` is a somewhat useful spec function (it is a sort of inductive constructor of binary `Nat`), but it could definitely have a longer name like `Nat.pushBit` or `Nat.shiftBit` or something along those lines.

`div2` we can just remove, there is no need for it in these days of kernel bignums.

Reply:
> I'm not a huge fan of `bit` in this namespace as is. The meaning isn't obvious to me and `Nat`. Could we replace `bit b n` with `2*n + ofBool b` where `abbrev ofBool b := cond b 1 0`?

There would be no need for this abbrev if there was a coercion from Bool to Nat. I'm in favor of such a coercion but there is room for debate.

Reply:
> Agreed, these definitions are holdovers from the lean 3 way of representing numbers, and I think we can just do without them. `bit` is a somewhat useful spec function (it is a sort of inductive constructor of binary `Nat`), but it could definitely have a longer name like `Nat.pushBit` or `Nat.shiftBit` or something along those lines.
> 
> `div2` we can just remove, there is no need for it in these days of kernel bignums.

I totally agree but there should be some more exposure of the kernel bignum so that people could implement and reason about efficient functions in Lean. Ultimately, Lean should only rely on a bignum library that only implements Lean-verified algorithms!

Reply:
For big `n`, `n / 2` is still slower than `n >>> 1`, although they are often both relatively fast. Maybe we should let the compiler optimize it.

Reply:
Are you sure (have you profiled)? Keep in mind that `/` has a kernel implementation but `>>>` does not.

Reply:
```lean
def div2 (x : Nat) : Nat := x / 2
def div2' (x : Nat) : Nat := x >>> 1

def run_div2 (n : Nat) (x : Nat) : IO Unit := do
  let mut t := 0
  for _ in [0 : n] do
    t := div2 x

def run_div2' (n : Nat) (x : Nat) : IO Unit := do
  let mut t := 0
  for _ in [0 : n] do
    t := div2' x

def n := 10^4
def x := 11^1000000

#eval timeit "x/2" (run_div2 n x)
#eval timeit "x>>>1" (run_div2' n x)
```
I'm not sure if it's the right way. In this example, `x >>> 1` is always significantly faster when `x` is large.

Reply:
Thanks for the analysis!  I think if `x >>> 1` is faster than `x/2` then `x >>> 1` is good with me.  One could also implement it in the kernel if that leads to significant speedup.

How about renaming `bit` to `shiftBit` and making `shiftBit` an abbreviation so it is transparent to decision procedures?

I'd also favor reordering the arguments to `def shiftBit : Nat -> Bool -> Nat` so you have a visually consistent ordering (e.g., `shiftBit 0b101 true = 0b1011`).  That's a taste issue though.

Reply:
heh, you know that order is deliberate...

Reply:
Yes, I figured it was deliberate.  Lean has always had most-significant digit on the left in numeric literals, so I'm trying to be consistent with that.

That said, I'd recommend we merge with `bit` renamed to `shiftBit` and the definition made an abbreviation.

Reply:
Why an abbreviation? That has specific consequences and I'm not sure they are relevant or appropriate here.

Reply:
inline does make sense though.

Reply:
I guess we could see.  My thinking is `shiftBit` would be something one would want a linear arithmetic decision procedure to see through.  Maybe that's premature.

Reply:
I would rather we just taught said linear arithmetic decision procedure about shiftBit directly.

Reply:
I'm not quite sure how to implement a linear arithmetic decision procedure without decoding `shiftBit` into a polynomial at some point, but I'm happy to table that discussion.  If there's a concrete need for a change we could make that later.  It still needs to be renamed from `bit` to `shiftBit`.

Reply:
Yes, I'm saying that the linear arithmetic procedure would see `shiftBit` and apply a theorem saying it's a specific polynomial because it's `shiftBit` and it knows about this function just like it knows about `+` and `*`. It can do this without `shiftBit` needing to be an `abbrev`.

Reply:
If this consensus is to rename `bit ` to `shiftBit`, I think a PR to make the rename in mathlib first would reduce friction when bumping Std.

I'm not a huge fan of the `shiftBit` name; it doesn't really make it clear where the bit is being shifted; `pushLsb` seems clearer in this regard, at the expense of not including the word `bit`.

Reply:
Indeed, we should try to harmonize with `Mathlib.Init.Data.Nat.Bitwise` and `Mathlib.Data.Nat.Bitwise`. They currently use [`shiftLeft'`](https://github.com/leanprover-community/mathlib4/blob/8f163221eaac63ae5e4d71fe84e34d68c727899c/Mathlib/Init/Data/Nat/Bitwise.lean#L180C1-L180C1) for this (not that this is better name). There's a couple more awkward conflicts there such as `shiftLeft_eq` and `shiftLeft_eq'`.

Reply:
> I'm not a huge fan of the shiftBit name; it doesn't really make it clear where the bit is being shifted; pushLsb seems clearer in this regard, at the expense of not including the word bit.

I'm also not a big fan of `shiftBit`; my suggestion would be `concatBit` or `concatMsb`, following the suggestion to reorder the arguments so that the new bit is the second argument. The operation, with that order, is somewhat like `List.concat`, so the name would be consistent with that.

Reply:
`concatBit` seem misleading to me, as its meaning is only correct if you think of a bit-vector as a *big*-endian list of bits; for *little* endian, it would be `consBit`.